### PR TITLE
Add --wait option with 'kubectl logs -f' to wait for the first container to start running, before providing logs

### LINKED ISSUE #1227: Forcing `kubectl log` to wait when container is creating
**What would you like to be added**:

When we do `kubectl logs <pod> [-f]` on a container in creation, we get this error message:
```
Error from server (BadRequest): container "<container>" in pod "<pod>" is waiting to start: [ContainerCreating, PodInitializing]
```
In this situation, in general we retry to launch the `kubectl logs <pod> [-f]` until the container is created.

A nice feature would be to have a the `kubectl logs` command to block until the container is created, so we run this command only one time.

**Why is this needed**:
To simplify the life of the user.

### PR DESCRIPTION
**Pull request by ankritisachan**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
`kubectl logs -f` fails when no container is ready to provide logs.
```
$ kubectl logs -f test -c nginx
Error from server (BadRequest): container "nginx" in pod "test" is waiting to start: CreateContainerConfigError
```
This PR adds a new flag `-F` to `kubectl logs`. This flag is supposed to follow the semantics of `tail -F`
Via this PR, if `-F` flag is used, 'kubectl logs` does not exit with 'Error from server...' message, instead waits for the first/all container/s to be running and then starts giving logs.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1227, https://github.com/kubernetes/kubernetes/issues/79547

#### Special notes for your reviewer:
The new flag '-F' waits for the container to be running before providing logs.

When used with '-p', i.e., '-p -F', it prints the previous logs and then continues streaming current logs. Currently '-p -f' behaves similar to '-p'. Hence, the combination of '-p -F' is implemented such that two API calls are made - first with '-p' and second with '-f'.

Along with the primary changes, I have moved `getContainerStatusByName` function out of 'staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go' to 'staging/src/k8s.io/kubectl/pkg/cmd/util/podcmd/podcmd.go', for easy reusability in other files. I, however, cannot move `waitForContainer`, since it performs some extra checks specific to flags used with `kubectl debug` in 'staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go'.

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -5952,6 +5952,10 @@ type PodLogOptions struct {
 	// the actual log data coming from the real kubelet).
 	// +optional
 	InsecureSkipTLSVerifyBackend bool `json:"insecureSkipTLSVerifyBackend,omitempty" protobuf:"varint,9,opt,name=insecureSkipTLSVerifyBackend"`
+
+	// wait for the container to start running before streaming logs. Defaults to false.
+	// +optional
+	Wait bool `json:"wait,omitempty" protobuf:"varint,10,opt,name=wait"`
```
Comment (dims) [inline staging/src/k8s.io/api/core/v1/types.go:5958]: kubectl changes should not be touching API. Is this really needed?

Comment (ankritisachan) [reply staging/src/k8s.io/api/core/v1/types.go]: Thanks for the review @dims! Fixed it with the latest changes.

**Comment by dims** [comment]
@ankritisachan the linter jobs fail on the same thing and easy to fix!

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go [hunk 2]
```diff
@@ -68,6 +68,9 @@ var (
 		# Begin streaming the logs of the ruby container in pod web-1
 		kubectl logs -f -c ruby web-1
 
+		# Wait for atleast one container to be ready before streaming the logs
+		kubectl logs -f web-1 -w
```
Comment (dims) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:72]: @ankritisachan so here let's look at `-c ruby` above, we should allow that firstly. Namely wait for the specified container and stream when it is ready.

second, if it is not specified, we should wait for ALL containers to be ready and not just one of them. 

(please think about the least surprise and best consistency and you will see why i am asking for this)

**Comment by apelisse** [comment]
the departure from `tail -f` is noticeable, and I would have expected the `-F` semantics to be implemented here. Here's the relevant section from `man tail`:
> ```
>      -F      The -F option implies the -f option, but tail will also check to see if the file being followed has
>             been renamed or rotated.  The file is closed and reopened when tail detects that the filename being
>             read from has a new inode number.
>
>             If the file being followed does not (yet) exist or if it is removed, tail will keep looking and will
>             display the file from the beginning if and when it is created.
>
>             The -F option is the same as the -f option if reading from standard input rather than a file.
> ```

Note that `tail -f /unknown-file` will hang until `/unknown-file` can be tailed, which is the behavior we're looking for here.
I'd try to stay as close as possible to this. If the container is restarted, I would also continue the logs, the same way `tail -F` would work (also please consider `kubectl logs --previous`).

A few more questions that needs an answer:
1. Should we be able to wait for a pod even if the pod doesn't exist yet?
2. If so, do we know which containers are part of the pod when we start waiting?
3. If we don't know which containers are part of the pod, how do we know if `-c` is needed or not (when `--all-containers` is false, `kubectl logs` requires `-c` if there's more than one pod).
4. When should we start printing the logs? For both variants of`--all-containers`

I think I have an opinion about all of these actually. 

I don't know exactly the extent of what @dims implied with his comment, but if you have `--all-containers=true`, I think we should start printing logs as soon as the first container prints logs. If we're looking for a specific container, then we should start printing logs as soon as that container is ready. I don't think we should allow doing `kubectl logs` for a pod that doesn't exist, which means we will know how many containers there are and we can enforce the required `-c` if there are more than one container.

**Comment by dims** [comment]
> I don't know exactly the extent of what @dims implied with his comment

+1 to what @apelisse said

**Comment by ankritisachan** [comment]
Hi @dims and @apelisse, Thanks for your comments.

I wanted to confirm my understanding here. We need the following conditions to be satisfied:

1. If --all-containers is true, we should print logs of the first container - I believe the existing implementation is already such that in case of multiple containers, we use a for loop to iterate over them and then call a function to print logs for each of them. Currently -f prints logs of the defaulted/first container only. I have just added a check before the printing happens which waits for the defaulted/first container to be running. So, I guess the default behaviour remains the same even after this check. Please let me know if my understanding is incorrect here, and I should implement anything additional for this case. Also, currently, -f streams the log of only the default container and not of the remaining container. Is this behaviour intentional?

2. If --all-containers is not specified we still wait for the default container and stream logs from it when it is running

3.  If a pod does not exist, throw an error and end - This is already taken care by kubectl logs' existing implementation. Do we want to change this behaviour where we keep waiting if pod does the exist? How long should we wait here?

4. If -c is used to specify the name of a container, wait for that container to get ready - This is also part of the existing implementation. My changes just wait for this container to get running and then starts streaming logs. 

5. If the specified container is not found, throw an error and end - This is the existing behavior. Do we want to change this behavior where we keep waiting for the container until it exists? How long should we wait here?

6. If no container is specified or --all-containers is false, we'll still stream logs from the defaulted container only - Again, this is the existing behavior. No changes made to this.

7. If logs are currently being streamed and in parallel if I delete and/or recreate a pod while the logs are streamed, the logs' streaming ends without any error. Is this expected? If not, what is the expected behavior?

Did I miss any scenario here? Please let me know your thoughts.

**Comment by apelisse** [comment]
> 1. If --all-containers is true, we should print logs of the first container - I believe the existing implementation is already such that in case of multiple containers, we use a for loop to iterate over them and then call a function to print logs for each of them. Currently -f prints logs of the defaulted/first container only. I have just added a check before the printing happens which waits for the defaulted/first container to be running. So, I guess the default behaviour remains the same even after this check. Please let me know if my understanding is incorrect here, and I should implement anything additional for this case. Also, currently, -f streams the log of only the default container and not of the remaining container. Is this behaviour intentional?

If all-containers only shows logs for one container, I'd argue that's a bug. If we can't follow the logs from all the files, then I'd say, we need to fail early in the command by saying the options are incompatible.

> 2. If --all-containers is not specified we still wait for the default container and stream logs from it when it is running

Assuming there's no `-c`, and there's only one container (is that what you call the default?), yes.

> 3. If a pod does not exist, throw an error and end - This is already taken care by kubectl logs' existing implementation. Do we want to change this behaviour where we keep waiting if pod does the exist? How long should we wait here?

I think the current behavior is fine.

> 4. If -c is used to specify the name of a container, wait for that container to get ready - This is also part of the existing implementation. My changes just wait for this container to get running and then starts streaming logs.

Sounds good to me!

> 5. If the specified container is not found, throw an error and end - This is the existing behavior. Do we want to change this behavior where we keep waiting for the container until it exists? How long should we wait here?

Current behavior sounds good.

> 6. If no container is specified or --all-containers is false, we'll still stream logs from the defaulted container only - Again, this is the existing behavior. No changes made to this.

Yes, just keep the existing behavior (which IIRC, fails if there's more than one). But just continue doing what we're doing.

> 7. If logs are currently being streamed and in parallel if I delete and/or recreate a pod while the logs are streamed, the logs' streaming ends without any error. Is this expected? If not, what is the expected behavior?

Yeah, that's expected. 

> Did I miss any scenario here? Please let me know your thoughts.

8. If the container is restarted, don't leave but resume printing logs from the new container
9. Consider logs from previous runs of the container

I think I'm still missing one thing, but that's probably good for now.

**Comment by ankritisachan** [comment]
> 9. Consider logs from previous runs of the container

Hi @apelisse, can you please elaborate this a bit? Do you mean to support the use of --wait with the previous option too? As in, keep waiting if we use `kubectl logs -p -w` until the container is killed and now we have some logs that are 'previous' in nature and then print those logs? Thank you!

**Comment by apelisse** [comment]
> Hi @apelisse, can you please elaborate this a bit? Do you mean to support the use of --wait with the previous option too? As in, keep waiting if we use kubectl logs -p -w until the container is killed and now we have some logs that are 'previous' in nature and then print those logs? Thank you!

Well, first I really don't think we should call it `-w/--wait` since it should be the semantics of -F. I guess  `-F` doesn't show you the content of previous files (since they've been overriden) so we should probably not do that. But if `-p` is used along with `-F` then yeah, I think we should print the previous logs, then the current logs, and then continue with the new logs if the container is restarted.

**Comment by ankritisachan** [comment]
> > 1. If --all-containers is true, we should print logs of the first container - I believe the existing implementation is already such that in case of multiple containers, we use a for loop to iterate over them and then call a function to print logs for each of them. Currently -f prints logs of the defaulted/first container only. I have just added a check before the printing happens which waits for the defaulted/first container to be running. So, I guess the default behaviour remains the same even after this check. Please let me know if my understanding is incorrect here, and I should implement anything additional for this case. Also, currently, -f streams the log of only the default container and not of the remaining container. Is this behaviour intentional?
> 
> If all-containers only shows logs for one container, I'd argue that's a bug. If we can't follow the logs from all the files, then I'd say, we need to fail early in the command by saying the options are incompatible.

Thanks for pointing that out. Apologies, I was incorrect about this. I verified again and --all-containers=true gives logs from all the containers.

**Comment by ankritisachan** [comment]
> > Hi @apelisse, can you please elaborate this a bit? Do you mean to support the use of --wait with the previous option too? As in, keep waiting if we use kubectl logs -p -w until the container is killed and now we have some logs that are 'previous' in nature and then print those logs? Thank you!
> 
> Well, first I really don't think we should call it `-w/--wait` since it should be the semantics of `-F`. I guess `-F` doesn't show you the content of previous files (since they've been overriden) so we should probably not do that. But if `-p` is used along with `-F` then yeah, I think we should print the previous logs, then the current logs, and then continue with the new logs if the container is restarted.

1. Do you mean, we should add `-F` instead of `-w/--wait` for this change?
2. Will `-F`(caps) not be confusing with `-f/--follow`(smallcase) that we already have? Also, if answer to the previous question is yes, will the usage be like `kubectl logs -f -F` or `kubectl logs -F` where we'll wait and stream current and future logs? Can you clarify more on the way new behaviour should be implemented? Do you intend to make `-f' and new `-F` mutually exclusive? If `-F` and advanced version of `-f`? I was implementing keeping in mind that `-f` remains as it is, but when we use `-w` the terminal keeps waiting for something to be available and then continue with the original `-f`.
3. Currently `kubectl logs -p -f` prints previous logs and exit. Are you suggesting when we use `kubectl logs -p -f -F`, we wait for previous logs to be available and then print previous, current, and continue streaming? Will this not be overlapping with the behaviour of `kubectl logs -f -F` except that it has the extra 'previous' logs?

**Comment by apelisse** [comment]
I think I see where some of the confusion comes from. I think the behavior of `kubectl logs -f` follows the same principles as `tail -f`. So I think it only makes sense to use the same logic as `tail -F` (see [tail man page](https://linux.die.net/man/1/tail)) for `kubectl logs`.

**Comment by dims** [comment]
@ankritisachan please fix the linter job. I don't think the other one is related. Also we don't have a QA team :) so do the best you can. See if you can rope in some folks/friends to kick the tires of this new option.

**Review by apelisse — COMMENTED** [review]
Unfortunately we're getting close to code-freeze and I have other features that have more urgency right now. I'd really love to review this but I don't know if I'll have the time to do it seriously. I'll try though.

**Comment by dims** [comment]
@apelisse ack, let me add milestone for 1.29, we can remove milestone if it gets too much

/milestone v1.29

**Comment by hailkomputer** [comment]
Hello, bug triage team lead here. I want to check the status and remind that the the code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03), as we want to ensure that each PR has a chance to be merged.
Are we still targeting the 1.29 release?

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go [hunk 3]
```diff
@@ -168,6 +186,7 @@ func NewCmdLogs(f cmdutil.Factory, streams genericiooptions.IOStreams) *cobra.Co
 func (o *LogsOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&o.AllContainers, "all-containers", o.AllContainers, "Get all containers' logs in the pod(s).")
 	cmd.Flags().BoolVarP(&o.Follow, "follow", "f", o.Follow, "Specify if the logs should be streamed.")
+	cmd.Flags().BoolVarP(&o.FollowWithWait, "follow-with-wait", "F", o.FollowWithWait, "Wait for the container to be in running state when the logs are streaming")
```
[the remark this answers is not in the document]
Comment (ankritisachan) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go]: Renamed 'FollowWithWait' to 'FollowWithRetry' and added a new option '--retry' that is allowed only with '--follow'.

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go [hunk 4]
```diff
@@ -168,6 +187,8 @@ func NewCmdLogs(f cmdutil.Factory, streams genericiooptions.IOStreams) *cobra.Co
 func (o *LogsOptions) AddFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&o.AllContainers, "all-containers", o.AllContainers, "Get all containers' logs in the pod(s).")
 	cmd.Flags().BoolVarP(&o.Follow, "follow", "f", o.Follow, "Specify if the logs should be streamed.")
+	cmd.Flags().BoolVar(&o.Retry, "retry", o.Retry, "Retry obtaining logs until available.")
+	cmd.Flags().BoolVarP(&o.FollowWithRetry, "", "F", o.FollowWithRetry, "Same as --follow --retry")
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:191]: that's perfect.

**Comment by apelisse** [comment]
Hopefully I'll have time to look at it deeply. I just reviewed the API and how the flags work and that looks absolutely great! Very happy with it. I'll take a look at how this actually behaves later :-)

**Comment by apelisse** [comment]
In the mean-time, please fix the verify script failure.
> ```
>  1. command "kubectl logs": flag name "" is invalid, long form of flag names can only contain lowercase characters or dash (must match ^[a-z]+[a-z\-]*$)
> ```

Looks like we have something wrong with the way the flag is built. Can we make a long-form only in cobra?

**Comment by apelisse** [comment]
We might want to change this regex/error: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/util/sanity/cmd_sanity.go#L139-L149
cc @fabianofranz

**Comment by ankritisachan** [comment]
> In the mean-time, please fix the verify script failure.
> 
> > ```
> >  1. command "kubectl logs": flag name "" is invalid, long form of flag names can only contain lowercase characters or dash (must match ^[a-z]+[a-z\-]*$)
> > ```
> 
> Looks like we have something wrong with the way the flag is built. Can we make a short-form only in cobra?

Hi @apelisse, I looked around and found that Cobra doesn't support using only shorthand yet. There were [attempts](https://github.com/spf13/pflag/pull/256) made but none of them got merged to the main spf13/pflag branch. 

One of the suggestions that were made (when I went through the discussions on the issue/PR) was to use the same text for long-form and short-form. Should I add that change? 
Or should I still update the regex as you have suggested?

**Comment by apelisse** [comment]
Or alternatively, you could name it `--follow-with-retry`, or merely `--retry` and have the logic do the right thing?

**Comment by ankritisachan** [comment]
> --follow-with-retry

Added --follow-with-retry to go with -F.

**Comment by dims** [comment]
/milestone clear

( too late for 1.29 :( since we missed code freeze let's try for 1.30! )

**Comment by apelisse** [comment]
To be honest, I'm happy with the API at that point, I'd appreciate someone else doing the technical review.

**Comment by npolshakova** [comment]
Hi @ankritisachan, @dims suggested I take a look at this PR as a way to learn and do some testing! I've tried it out locally and had some questions. 

I first tested it with a test curl deployment and it worked as expected with a pod with a single container and a pod with a sidecar injected 🎉 

```
> _output/bin/kubectl get pods -n curl
NAME                    READY   STATUS    RESTARTS   AGE
curl-5c78c847cd-hsv9c   1/1     Running   0          104m
```

Not providing the container works as expected with `kubectl logs -F deployment/curl -n curl`.


```
_output/bin/kubectl get pods -n curl-inmesh
NAME                    READY   STATUS             RESTARTS      AGE
curl-56d5bb7d94-5lssx   1/2     ImagePullBackOff   0             71m
curl-5c78c847cd-p5tg8   2/2     Running            2 (63m ago)   74m
```

Where I broke one of the curl containers with an image pull error and saw the expected difference in behavior between `-f` and `-F`:

```
❯ _output/bin/kubectl logs -f curl-56d5bb7d94-5lssx -n curl-inmesh -c curl
Error from server (BadRequest): container "curl" in pod "curl-56d5bb7d94-5lssx" is waiting to start: image can't be pulled


# No error here with -F flag:
❯ _output/bin/kubectl logs -F curl-56d5bb7d94-5lssx -n curl-inmesh -c curl

```

Where I can get the logs from the istio-proxy container as expected even when the curl container is in an error state:

```
❯ _output/bin/kubectl logs -F curl-56d5bb7d94-5lssx -n curl-inmesh -c istio-proxy
2023-11-25T17:12:56.654627Z	info	FLAG: --concurrency="0"
2023-11-25T17:12:56.654643Z	info	FLAG: --domain="curl-inmesh.svc.cluster.local"
2023-11-25T17:12:56.654646Z	info	FLAG: --help="false"
2023-11-25T17:12:56.654648Z	info	FLAG: --log_as_json="false"
2023-11-25T17:12:56.654649Z	info	FLAG: --log_caller=""
2023-11-25T17:12:56.654650Z	info	FLAG: --log_output_level="default:info"
```

When I tested a Job with a [pod failure policy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy), I'm seeing slightly different behavior between `-f` and `-F`. 

When I use `-f`:

``` 
❯ _output/bin/kubectl logs -f jobs/job-pod-failure-policy-example
Found 2 pods, using pod/job-pod-failure-policy-example-d7jv9
Hello world!

❯ _output/bin/kubectl logs -f pod/job-pod-failure-policy-example-d7jv9
Hello world!
```

Versus when I use `-F`, the previous logs are not shown and no new logs appear: 

``` 
❯ _output/bin/kubectl logs -F  jobs/job-pod-failure-policy-example
Found 2 pods, using pod/job-pod-failure-policy-example-d7jv9


❯ _output/bin/kubectl logs -F pod/job-pod-failure-policy-example-d7jv9

# The --previous flag does return logs either since container is not restarted
❯ _output/bin/kubectl logs -pF  jobs/job-pod-failure-policy-example
Found 3 pods, using pod/job-pod-failure-policy-example-d7jv9
Error from server (BadRequest): previous terminated container "main" in pod "job-pod-failure-policy-example-d7jv9" not found
```

If I look at the logs for the pods, I see the job is (failing) as expected: 

``` 
❯ _output/bin/kubectl get pods
NAME                                   READY   STATUS      RESTARTS   AGE
job-pod-failure-policy-example-d7jv9   0/1     Error       0          2m30s
job-pod-failure-policy-example-s9x9z   0/1     Error       0          2m30s
job-pod-failure-policy-example-zljxr   0/1     Error       0          2m30s
```

If I look at the logs with timestamps I see the expected logs:

```
❯ kubectl logs -f --selector job-name=job-pod-failure-policy-example --timestamps
2023-11-25T18:34:41.482258209Z job-pod-failure-policy-example-zljxr Hello world!
2023-11-25T18:34:41.482562876Z job-pod-failure-policy-example-s9x9z Hello world!
2023-11-25T18:34:41.481397584Z job-pod-failure-policy-example-d7jv9 Hello world!

```

But the `-F` flag does not pick up these logs. Is that expected behavior for Jobs? Thanks!

**Comment by apelisse** [comment]
Woa, very good job on reproducing this! This shows that we need more tests to assert that everything is working properly. Thanks!

**Comment by rayjlinden** [comment]
What happened to this PR?  Could use this functionality...

**Pull request closed without merging**